### DISCUSSION TITLE
/hyndsight/ts2csv/

### DISCUSSION THREAD

**Discussion by robjhyndman** [robjhyndman/robjhyndman.com#298] [Announcements]
Saving ts objects as csv files 
 https://robjhyndman.com/hyndsight/ts2csv/

**Comment by robjhyndman** [comment]
```header
Originally posted by Ícaro Agostino on 2018-08-07 12:13:31
````
very useful, thanks professor!

**Comment by robjhyndman** [comment]
```header
Originally posted by Earo Wang on 2018-08-11 05:59:34
````
Hi Rob,

\`if-else\` can be replaced by one liner:

ts2csv \&lt;- function(x) {\
fname \&lt;- paste0(deparse(substitute(x)), &quot;.csv&quot;)\
readr::write_csv(tsibble::as_tsibble(x, gather = FALSE), fname)\
}

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-08-11 06:54:28
````
This is why you\'re the main tsibble developer and not me. :) Thanks!
I\'ll update the post with your code.

**Comment by PhilBla** [comment]
Hi Rob.

I must thank you for your time spent on creating tools and educating in their use.  I am fairly new to R and time series, and my accounting end users would like their output in excel.

Your function above is great for csv output but are you able to help with a similar export to excel?

So
 readr::write_csv(tsibble::as_tsibble(visnights), "visnights.csv") 
works fine (after removing the gather = FALSE) with an index column containing "1998 Q1" as its first entry

but if I use something like 
write_xlsx(tsibble::as_tsibble(visnights), path = "visnights.xlsx", col_names = TRUE)
The index column first entry is 10227 - which I believe is the number of days after 1 January 1970

Any suggestions on which package to use to write to excel with a "1988 Q1" formatted index ?

Many thanks
Philip

**Reply by PhilBla** [reply]
I'm not sure if this is monitored, and didn't want to cross\double post, so please note that I have posted the question on stackoverflow 

https://stackoverflow.com/questions/77126031/how-can-i-export-a-tsibble-to-excel-retaining-the-time-index-as-character-forma

**Reply by robjhyndman** [reply]
It looks like you now have an answer on SO

**Reply by PhilBla** [reply]
Thanks

**Comment by ppapanastasiou** [comment]
Note that tsibble::as_tsibble() function no longer supports the gather argument, which was deprecated in earlier versions of the tsibble package.  You will need to use this version of the function 

################################################################################
# Convert time series data to CSV
# Compatible with both univariate and multivariate 'ts' objects
# Works with fpp2-style datasets like auscafe, usconsumption, etc.
################################################################################

ts2csv <- function(x) {
  fname <- paste0(deparse(substitute(x)), ".csv")

  if (!is.ts(x)) {
    stop("Input must be a time series object of class 'ts'.")
  }

  time_index <- time(x)

  if (is.null(dim(x))) {
    # Univariate time series
    df <- tibble::tibble(
      time = as.Date(as.yearmon(time_index)),
      value = as.numeric(x)
    )
  } else {
    # Multivariate time series
    df <- tibble::tibble(
      time = as.Date(as.yearmon(time_index))
    )
    df <- cbind(df, as.data.frame(x))
  }

  readr::write_csv(df, fname)
  message("Saved to ", fname)
}

# Example usage
library(fpp2)
library(tsibble)

# Save univariate time series to CSV
ts2csv(auscafe)

ts2csv(visnights)  # Multivariate quarterly data
ts2csv(elecdemand) # Multivariate half-hourly data

**Reply by robjhyndman** [reply]
Thanks. Now fixed by using the `pivot_longer` argument instead.
